consensus/fork_choice/src/fork_choice.rs

```diff
@@ -1647,10 +1682,38 @@ where
 /// Helper struct that is used to encode/decode the state of the `ForkChoice` as SSZ bytes.
 ///
 /// This is used when persisting the state of the fork choice to disk.
-#[derive(Encode, Decode, Clone)]
+
+pub type PersistedForkChoice = PersistedForkChoiceV20;
+
+#[superstruct(
+    variants(V19, V20),
```

Comment:
According to the versioning scheme we've got outlined here, the struct should be named based on the version it was introduced, so I think we want `variants(V1, V20)` here since `PersistedForkChoice` has always existed: https://github.com/sigp/lighthouse/blob/stable/beacon_node/beacon_chain/src/schema_change/README.md#naming-conventions

Reply:
I asked @michaelsproul earlier about versioning this inner `PersistedForkChoice` because I noticed we had a separate [outer `PersistedForkChoice`](https://github.com/sigp/lighthouse/blob/v4.6.0/beacon_node/beacon_chain/src/persisted_fork_choice.rs#L15) and I was worried that the outer one was created to shield the inner one from versioning or something. He told me we've already had different versions of the inner `PersistedForkChoice`. They just get removed eventually as we stop supporting super old DB versions.

So it wouldn't make sense to name it `V1` since it hasn't actually been the same since `V1`. I suppose we *could* look at old code to try and figure out the last time this struct was modified. But before we go there let's back up & consider this comment from the guide you linked (which I'm going to slightly edit to adhere to the discussion at hand):

> Previously the schema migration code would name types by the last version at which they were valid. For example if `PersistentForkChoice` changed in `V20` then we would name the two variants `PersistentForkChoiceV19` and `PersistentForkChoiceV20`. The problem with this scheme is that if `PersistentForkChoice` changes again in the future at say v24 then `PersistentForkChoiceV20` would need to be renamed to `PersistentForkChoiceV23`, which is annoying. Using the first valid version as described above does not have this issue.

I agree that being forced to rename is very annoying. But there's actually no reason we *have to* rename `PersistentForkChoiceV20` to `PersistentForkChoiceV23`. We could just as easily carry on with `PersistentForkChoiceV19`,`PersistentForkChoiceV20`, and `PersistentForkChoiceV24`. After all, these structs are still **named after the versions in which they are used**.

The only issue is communicating this easily in naming conventions guide.

Perhaps instead of:
> Prefer to name versions of structs by the version at which the change was introduced. 

we might write something along the lines of:

> When introducing new versions of a struct, adhere to the following convention:
>
> If the struct has no previous versions in the codebase, the original version should be named by the *last* version in which it was valid. All subsequent versions should be named after the version in which they were introduced. For example, if the struct `Foo` has no previous migrations, and you added a field to `Foo` in v9, then you would call the original version `FooV8` and write a schema change that migrates from `FooV8` to `FooV9`. If at some later date, another field is added to `Foo` in v12, you would add a schema change that migrates from `FooV9` to `FooV12`.

Reply:
While we're on the subject, I had another thought for the schema migrations guide. We should encourage the developer to add `impl Into<FooV9> for FooV8` and `impl Into<FooV8> for FooV9` just below the definition of the struct. Note that this encourages them to **be explicit about which versions are being transformed**.

Implementing this adds a significant amount of clarity about how versions are migrated and which versions are transformed into which. I previously got tripped up about this when I didn't implement these and introduced bugs in the migration code. You can leverage these `impl`s in the schema migration code to make that simple as well. Also by being explicit about the versions you avoid [having to clutter your diff with these lines](https://github.com/sigp/lighthouse/pull/5128/files#diff-c1f445f57347fe0e25d08edb377cd5ae462a6ad9e7addb1b915208680f740b6bL23) when adding a new version. I already started to adopt this convention which is the only reason [I made changes to `beacon_fork_choice_store.rs`](https://github.com/sigp/lighthouse/pull/5128/files#diff-e067003ed49b9518588f2128b05809057ab223c630c1b7dc4e7e2ae29e0076a2L392)

Reply:
> He told me we've already had different versions of the inner PersistedForkChoice. They just get removed eventually as we stop supporting super old DB versions.

Right that would make it tough to track introduced version. I like the suggestion to use the previous version if there are no versions defined in the super struct. 

I'm on board with what you're suggesting re migration guide. And explicit intos as well